BLS2-65903-event-organization

Zhltsk-V opened the pull request.

Commit:
Реализована возможность создавать(проводить) событие.

Commit:
Task BJS2-65903-event-organization completed.

marsel-mkh commented on src/main/java/school/faang/user_service/dto/event/EventDto.java:
```diff
@@ -0,0 +1,23 @@
+    private LocalDateTime endDate;
+    private Long ownerId;
+    private String description;
+    private List<Long> relatedSkills;
```
Как понимаю, List<Long> relatedSkills это лист айди скилов, если так то можно переименовать в relatedSkillsId

marsel-mkh commented on src/main/java/school/faang/user_service/controller/event/EventController.java:
```diff
@@ -0,0 +1,58 @@
+        return eventService.create(event);
+    }
+
+    private void validateEvent(EventDto event) {
```
Валидацию можно вынести ниже основных методов, будет чуть элегантнее.
По валидации:
давай вынесем event.getTitle() == null в отдельный if и выкинем исключение с текстом " Event title cannot be null."
Давай добавим логирование в контроллер, так будет удобнее при багфиксе искать ошибки.

marsel-mkh commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,120 @@
+    private final SkillRepository skillRepository;
+    private final EventMapper eventMapper;
+
+    public EventDto create(@NotNull(message = "Event can not be null") EventDto eventDto) {
```
Валидацию @NotNull(message = "Event can not be null") можно вынести в контроллер

marsel-mkh replied:
![image](https://github.com/user-attachments/assets/d2133c3a-02b2-40f7-a285-7cabcf7c6647)

marsel-mkh commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    private void validateUserSkills(@NotNull(message = "Event can not be null") EventDto eventDto) {
+        List<Long> requiredSkills = eventDto.getRelatedSkills();
```
Поле relatedSkills в эвенте выглядит так:     private List<Skill> relatedSkills;
в дто же ты получаешь только айди:     private List<Long> relatedSkills;
так что тут List<Long> requiredSkills это тоже лист айди. 
 В данном случае нужно через репозиторий получать все скилы (findAllByID()) и сравнивать их

marsel-mkh replied:
if (requiredSkills == null || requiredSkills.isEmpty()) нужно вынести в отдельный блок кода, либо же добавить else. Если у тебя скилы  не проходят if (requiredSkills != null && !requiredSkills.isEmpty()) , то метод валидации бесполезен.

marsel-mkh commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,120 @@
+    public EventDto create(@NotNull(message = "Event can not be null") EventDto eventDto) {
+        validateUserSkills(eventDto);
+        Event event = eventMapper.toEntity(eventDto);
+        event.setOwner(getUserById(eventDto.getOwnerId()));
```
Давай вынесем getUserById(eventDto.getOwnerId() и mapSkillIdsToEntities(eventDto.getRelatedSkills() в отдельные переменные, так будет читабельнее

marsel-mkh commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,120 @@
+        }
+    }
+
+    private User getUserById(Long userId) {
```
Если создаешь такой метод, то его можно использовать и на 40 строке тоже

marsel-mkh replied:
Доп методы лучше вынести ниже, будет более читабельно

marsel-mkh commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,120 @@
+
+    private List<Skill> mapSkillIdsToEntities(@NotNull(message = "Skills can not be null") List<Long> skillIds) {
+        return skillIds.stream()
+                .map(id -> skillRepository.findById(id)
```
skillRepository::findById

marsel-mkh commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,120 @@
+        return skillIds.stream()
+                .map(id -> skillRepository.findById(id)
+                        .orElseThrow(() -> new DataValidationException("Skill not found for id: " + id)))
+                .collect(Collectors.toList());
```
лучше использовать .toList(); , чем .collect(Collectors.toList());

marsel-mkh commented on src/main/java/school/faang/user_service/controller/event/EventController.java:
```diff
@@ -0,0 +1,58 @@
+        return eventService.getEvent(eventId);
+    }
+
+    public List<EventDto> getEventService(EventFilterDto filter) {
```
Название у метода интересное, наверное должно быть getEventsByFilter(EventFilterDto filter)

marsel-mkh commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,120 @@
+
+    public List<EventDto> getEventsByFilter(@NotNull(message = "Event can not be null") EventFilterDto filter) {
+        List<Event> allEvents = eventRepository.findAll();
+        List<Event> eventsAfterFilter = allEvents.stream()
```
Можно вынести фильтры в отдельный метод и подумать что то с лямбдами, чтоб соблюдать open close principe

marsel-mkh commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,120 @@
+        List<Event> allEvents = eventRepository.findAll();
+        List<Event> eventsAfterFilter = allEvents.stream()
+                .filter(event -> filter.getTitle() == null || event.getTitle()
+                        .contains(filter.getTitle()))
```
Хорошо реализовал проверку, молодец!

marsel-mkh commented on src/main/java/school/faang/user_service/controller/event/EventController.java:
```diff
@@ -0,0 +1,58 @@
+        return eventService.getEventsByFilter(filter);
+    }
+
+    public void deleteEvent(long eventId) {
```
Можно возвращать булеан

marsel-mkh commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    public void deleteEvent(long eventId) {
+        eventRepository.findById(eventId)
```
Для проверки наличия эвента есть метод existById(), можно проверять через него

RinatBeybutov replied:
этот комментарий не пофикшен. Есть разница между findById и existById

marsel-mkh commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,120 @@
+        return eventMapper.toDto(event);
+    }
+
+    private void validation(@NotNull(message = "Event can not be null") Event event) {
```
Хорошо сделал валидацию!

marsel-mkh commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,120 @@
+                .map(eventMapper::toDto)
+                .toList();
+    }
+}
```
в конце лишняя строка

marsel-mkh commented on src/main/java/school/faang/user_service/dto/event/EventDto.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.List;
+
+@Data
+public class EventDto {
```
![image](https://github.com/user-attachments/assets/f6a56526-cffa-4fc2-865f-27cb690e0213)
можно сделать так

marsel-mkh commented on src/main/java/school/faang/user_service/mapper/EventMapper.java:
```diff
@@ -0,0 +1,37 @@
+    Event toEntity(EventDto eventDto);
+
+    @Mapping(target = "ownerId", source = "owner.id")
+    @Mapping(target = "relatedSkills", source = "relatedSkills", qualifiedByName = "skillsToIds")
```
по идее у тебя там лист айдишников и можно сделать что то на подобии:     @Mapping(source = "receiver.id", target = "receiverId")
Насколько я знаю, маппер поддерживает маппинг листов таких образом

marsel-mkh replied:
Вариант с qualifiedByName тоже хорош!

Commit:
Changes have been made to task BJS2-65903-event-organization:

- Fixed the filtering logic
- Improved validation in the controller
- Minor issues have been corrected

Commit:
Changes have been made to task BJS2-65903-event-organization:

- Added logging to the service
- Fixed minor errors

Commit:
Changes have been made to task BJS2-65903-event-organization:

- added JavaDoc

Commit:
Changes have been made to task BJS2-65903-event-organization:

RinatBeybutov commented on src/main/java/school/faang/user_service/controller/event/EventController.java:
```diff
@@ -0,0 +1,116 @@
+     * @param filter DTO фильтра, содержащее критерии поиска.
+     * @return Список событий в виде {@link List<EventDto>}.
+     */
+    public List<EventDto> getEventsByFilter(@Valid EventFilterDto filter) {
```
можно просто getEvents

RinatBeybutov commented on src/main/java/school/faang/user_service/mapper/EventMapper.java:
```diff
@@ -0,0 +1,37 @@
+        }
+        return skills.stream()
+                .map(Skill::getId)
+                .collect(Collectors.toList());
```
просто toList

RinatBeybutov commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,243 @@
+     * @throws DataValidationException если пользователь не обладает необходимыми навыками.
+     */
+    public EventDto create(EventDto eventDto) {
+        log.info("Creating event with data: {}", eventDto);
```
думаю, что это уже перебор с логгированием. У тебя в контроллере уже стоят логи.
Ну или на худой конец поменять уровень логгирования на debug

RinatBeybutov commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,243 @@
+     * @return Список событий в виде {@link List<EventDto>}.
+     */
+    public List<EventDto> getEventsByFilter(EventFilterDto eventFilterDto) {
+        log.info("Fetching events by filter: {}", eventFilterDto);
```
здесь нужно предусмотреть ситуацию, когда fiter = null

RinatBeybutov commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,243 @@
+     * @param event Событие для валидации.
+     * @throws DataValidationException если пользователь не может проводить событие.
+     */
+    private void validation(Event event) {
```
название метода лучше сделать более значимым, например validateUserAccess или checkAccessToEvent. Подумай, может что лучше придумаешь 

Название метода должно быть глаголом, а не существительным

RinatBeybutov commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,243 @@
+        log.debug("Validating event: {}", event);
+        User owner = event.getOwner();
+        List<Skill> skillsOwner = owner.getSkills();
+        for (Skill skill : skillsOwner) {
```
это можно в стримы переписать, если есть желание. Можно и так оставить

RinatBeybutov commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,243 @@
+     * @param userId Идентификатор пользователя.
+     * @return Список событий в виде {@link List<EventDto>}.
+     */
+    public List<EventDto> getOwnerEvent(long userId) {
```
getUsersEvents. Если такое название было в задании, то ок

Zhltsk-V replied:
Создайте метод getOwnedEvents. Этот метод будет обрабатывать запрос и предоставлять список событий.

Так было в задании.

RinatBeybutov commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,243 @@
+        List<EventDto> eventDtos = events.stream()
+                .map(eventMapper::toDto)
+                .toList();
+        log.info("Found {} events for owner with ID: {}", eventDtos.size(), userId);
```
можно второе логгирование хотя бы убрать и в один стрим с возвращением написать

RinatBeybutov commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,243 @@
+        List<EventDto> eventDtos = events.stream()
+                .map(eventMapper::toDto)
+                .toList();
+        log.info("Found {} participated events for user with ID: {}", eventDtos.size(), userId);
```
можно второе логгирование хотя бы убрать и в один стрим с возвращением написать

RinatBeybutov commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,243 @@
+     */
+    private void validateUserSkills(EventDto eventDto) {
+        log.debug("Validating user skills for event: {}", eventDto);
+        List<Long> requiredSkillIds = eventDto.getRelatedSkillsId();
```
тут должно быть relatedSkillsIds вместо requiredSkillIds?

RinatBeybutov commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,243 @@
+            return;
+        }
+
+        User user = userRepository.findById(eventDto.getOwnerId())
```
здесь лучше заменить на метод existById

Commit:
Changes have been made to task BJS2-65903-event-organization:

- The logic with DTO has been modified (eventCreateDto and eventViewDto have been added)

- Minor issues have been fixed

Commit:
Changes have been made to task BJS2-65903-event-organization:
- Minor issues have been fixed

Commit:
Merge branch 'murloc-master-stream9' into feature/BJS2-65903-event-organization

Commit:
Changes have been made to task BJS2-65903-event-organization:
- Rename filter
